File: 20.valid-parentheses.js
/*
 * @lc app=leetcode id=20 lang=javascript
 *
 * [20] Valid Parentheses
 */

// @lc code=start
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function(s) {
  let isOpenArray = [];
  for (let i = 0; i < s.length; i++) {
    let isOpen = isOpenArray.pop();
    if (!isOpen && (s[i] === ')' || s[i] === ']' || s[i] === '}')) {
      return false;
    } else if (isOpen === '(' && (s[i] === '}' || s[i] === ']')) {
      return false;
    } else if (isOpen === '[' && (s[i] === ')' || s[i] === '}')) {
      return false;
    } else if (isOpen === '{' && (s[i] === ')' || s[i] === ']')) {
      return false;
    } else if (s[i] === '(' || s[i] === '[' || s[i] === '{') {
      if (isOpen) {
        isOpenArray.push(isOpen);
      }
      isOpenArray.push(s[i]);
    }
  }
  return isOpenArray.length === 0;
};
// @lc code=end
// 76/76 cases passed (48 ms)
// Your runtime beats 95.2 % of javascript submissions
// Your memory usage beats 10 % of javascript submissions (35.9 MB)

File: 20.valid-parentheses.java
import java.util.ArrayList;

/*
 * @lc app=leetcode id=20 lang=java
 *
 * [20] Valid Parentheses
 */

// @lc code=start
class Solution {
    public boolean isValid(String s) {
        ArrayList<String> isOpenArray = new ArrayList<String>();
        for (int i = 0; i < s.length(); i++) {
            String isOpen = "";
            if (!isOpenArray.isEmpty()) {
                isOpen = isOpenArray.remove(isOpenArray.size() - 1);
            }
            if (isOpen.equals("") && (s.charAt(i) == ')' || s.charAt(i) == ']' || s.charAt(i) == '}')) {
                return false;
            } else if (isOpen.equals("(") && (s.charAt(i) == '}' || s.charAt(i) == ']')) {
                // System.out.println(1);
                return false;
            } else if (isOpen.equals("[") && (s.charAt(i) == ')' || s.charAt(i) == '}')) {
                // System.out.println(2);
                return false;
            } else if (isOpen.equals("{") && (s.charAt(i) == ')' || s.charAt(i) == ']')) {
                // System.out.println(3);
                return false;
            } else if (s.charAt(i) == '(' || s.charAt(i) == '[' || s.charAt(i) == '{') {
                if (isOpen != "") {
                    isOpenArray.add(isOpen);
                }
                isOpenArray.add(String.valueOf(s.charAt(i)));
            }
        }
        // System.out.println(4);
        return isOpenArray.isEmpty();
    }
}
// @lc code=end
// 76/76 cases passed (2 ms)
// Your runtime beats 62.15 % of java submissions
// Your memory usage beats 100 % of java submissions (34.4 MB)